src/senaite/core/z3cform/widgets/duration/widget.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of SENAITE.CORE.
+#
+# SENAITE.CORE is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free Software
+# Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright 2018-2024 by it's authors.
+# Some rights reserved, see README and LICENSE.
+
+from datetime import timedelta
+
+from bika.lims import api
+from senaite.core.interfaces import ISenaiteFormLayer
+from senaite.core.schema.interfaces import IDurationField
+from senaite.core.z3cform.interfaces import IDurationWidget
+from senaite.core.z3cform.widgets.basewidget import BaseWidget
+from z3c.form.browser import widget
+from z3c.form.browser.widget import HTMLInputWidget
+from z3c.form.converter import TimedeltaDataConverter
+from z3c.form.interfaces import IFieldWidget
+from z3c.form.widget import FieldWidget
+from zope.component import adapter
+from zope.interface import implementer
+
+
+@adapter(IDurationField, IDurationWidget)
+class DurationDataConverter(TimedeltaDataConverter):
+    """Converts the value between the field and the widget
+    """
+
+    def toWidgetValue(self, value):
+        """Converts from field value to widget
+        """
+        if not isinstance(value, timedelta):
+            return {}
+
+        # Note timedelta keeps days and seconds a part!
+        return {
+            "days": value.days,
+            "hours": (value.seconds / 3600) % 24, # hours within a day
+            "minutes": (value.seconds / 60) % 60, # minutes within an hour
+            "seconds": value.seconds % 60, # seconds within a minute
+        }
+
+    def toFieldValue(self, value):
+        """Converts from widget to field value
+        """
+        # days, hours, minutes, seconds
+        dhms = list(value) + [0]*4
```

Comment:
Is this one too much? I.e. `dhms = list(value) + [0]*3`?
But actually I do not understand the code here.
Why is this dance here ndeeded with `dhms`?
As far as I can see, the value is something like this: `{"days": 2, "hours": 48, "minutes": 70, "seconds": 70}`. Why the list conversion? It will just take the keys of the dict: `list({"days": 2, "hours": 48, "minutes": 70, "seconds": 70}) ->  ['days', 'hours', 'minutes', 'seconds']`

Reply:
I was getting the values as a list on submit. You right, way better with https://github.com/senaite/senaite.core/pull/2482/commits/925256bf2b8352ef4f2089b832379552fcd02ec1